Action container log collection does not wait for sentinel on developer error

sven-lange-last opened the pull request:
* So far, log collection for managed Docker action containers waits for sentinels - even if a developer error occurs. Only if an action exceeded its init or run timeout, sentinel waiting will be disabled.
* This change disables sentinel waiting for all developer errors because a managed action runtime may not be able to write sentinels at all in developer error situations. With this change, missing sentinels are handled in a more robust way.
* Without sentinel waiting, the Docker action container log file is only read until file end is reached. No re-reads are performed if the sentinel has not been found yet. On busy systems, this may lead to a few missing log lines in case of developer errors.

## Related issue and scope
<!--- Please include a link to a related issue if there is one. -->
- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components
<!--- Select below all system components are affected by your change. -->
<!--- Enter an `x` in all applicable boxes. -->
- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [x] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes
<!--- What types of changes does your code introduce? Use `x` in all the boxes that apply: -->
- [x] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
<!--- Please review the points below which help you make sure you've covered all aspects of the change you're making. -->

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

sven-lange-last commented:
## Assessment

* Following assessment lists all situations where a container / developer error may occur.
* This PR changes log collecting and processing for container / developer errors.
* While this PR makes log collecting and processing more robust when container / developer errors are detected, it has the downside that a few log lines may be lost.

## Conditions for container / developer error

### In `ContainerProxy`:

* https://github.com/apache/openwhisk/blob/9bef49fcd47f7922e4226ff7a97385f1313f7d64/core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerProxy.scala#L280

  * Application-specific error while starting a blackbox container.
  * Examples: image not found in registry or network failure while pulling.
  * The action container does not even start, no log reading is performed at all.

### In `Container`:

* https://github.com/apache/openwhisk/blob/9bef49fcd47f7922e4226ff7a97385f1313f7d64/common/scala/src/main/scala/org/apache/openwhisk/core/containerpool/Container.scala#L140

  * Action timeout is exceeded during init.
  * Example: a Node.js action with a busy loop during init blocks the event loop.
  * Expectation: managed action runtimes do not produce sentinels on timeout during init.

* https://github.com/apache/openwhisk/blob/9bef49fcd47f7922e4226ff7a97385f1313f7d64/common/scala/src/main/scala/org/apache/openwhisk/core/containerpool/Container.scala#L177

  * Action timeout is exceeded during run.
  * Example: action runs too long.
  * Expectation: managed action runtimes do not produce sentinels on timeout during run.

### In `ActivationResult` - `processInitResponseContent()`:

* https://github.com/apache/openwhisk/blob/9bef49fcd47f7922e4226ff7a97385f1313f7d64/common/scala/src/main/scala/org/apache/openwhisk/core/entity/ActivationResult.scala#L151

  * POST /init returns a response with status != OK (200), response is not truncated and is a valid JSON. We assume that the managed action runtime detected a problem during init and sent the response.
  * Example: Node.js action code that only contains comment => no valid entry point can be found.
  * Expectation: managed action runtimes should produce sentinels in this situation.

* https://github.com/apache/openwhisk/blob/9bef49fcd47f7922e4226ff7a97385f1313f7d64/common/scala/src/main/scala/org/apache/openwhisk/core/entity/ActivationResult.scala#L153

  * POST /init returns a response with status != OK (200), response is not truncated but is no valid JSON. We assume that the managed action runtime detected a problem during init and sent a broken response due to a bug.
  * Should not happen for managed runtimes because such failures should be caught in test. Blackbox containers may return improper responses. Can malicious actions cause sending of a broken response?
  * Example: no known examples.
  * Expectation: managed action runtimes should produce sentinels in this situation.

* https://github.com/apache/openwhisk/blob/9bef49fcd47f7922e4226ff7a97385f1313f7d64/common/scala/src/main/scala/org/apache/openwhisk/core/entity/ActivationResult.scala#L157

  * POST /init returns a response with status != OK (200) and response is truncated because it exceeds the size limit (1 MiB). We assume that the managed action runtime detected a problem during init and sent a broken response due to a bug.
  * Should not happen for managed runtimes because such failures should be caught in test. Blackbox containers may return improper responses. Can malicious actions cause sending of a broken response?
  * Example: no known examples.
  * Expectation: managed action runtimes should produce sentinels in this situation.

* https://github.com/apache/openwhisk/blob/9bef49fcd47f7922e4226ff7a97385f1313f7d64/common/scala/src/main/scala/org/apache/openwhisk/core/entity/ActivationResult.scala#L161

  * The TCP connection to the action container "fails" or breaks down during POST /init or while waiting for a response AND the OOM check is positive. We assume that the managed action runtime did not have a chance to react on this error situation.
  * Example: action code consumes too much physical memory by including a lot of prereq packages.
  * Expectation: managed action runtimes won't produce sentinels in this situation.

* https://github.com/apache/openwhisk/blob/9bef49fcd47f7922e4226ff7a97385f1313f7d64/common/scala/src/main/scala/org/apache/openwhisk/core/entity/ActivationResult.scala#L165

  * The TCP connection to the action container "fails" or breaks down during POST /init or while waiting for a response AND the OOM check is negative. We assume that the managed action runtime did not have a chance to react on this error situation.
  * Example: the action code breaks the managed action runtime process by exiting.
  * Expectation: managed action runtimes won't produce sentinels in this situation. Malicious actions may be able to delay / prevent sentinel production.

### In `ActivationResult` - `processRunResponseContent()`:

* https://github.com/apache/openwhisk/blob/9bef49fcd47f7922e4226ff7a97385f1313f7d64/common/scala/src/main/scala/org/apache/openwhisk/core/entity/ActivationResult.scala#L197

  * POST /run returns a response with status != OK (200), response is not truncated and is a valid JSON. We assume that the managed action runtime detected a problem during / after run and sent the response.
  * Example: action runtime caught an exception thrown by action code.
  * Expectation: managed action runtimes should produce sentinels in this situation.

* https://github.com/apache/openwhisk/blob/9bef49fcd47f7922e4226ff7a97385f1313f7d64/common/scala/src/main/scala/org/apache/openwhisk/core/entity/ActivationResult.scala#L202

  * POST /run returns a response with status != OK (200), response is not truncated and is a valid JSON - but JSON is not an object (instead: String, Number, ...). We assume that the managed action runtime detected a problem during / after run and and sent a broken response due to a bug.
  * Should not happen for managed runtimes because such failures should be caught in test. Blackbox containers may return improper responses. Can malicious actions cause sending of a broken response?
  * Example: no known examples.
  * Expectation: managed action runtimes should produce sentinels in this situation.

* https://github.com/apache/openwhisk/blob/9bef49fcd47f7922e4226ff7a97385f1313f7d64/common/scala/src/main/scala/org/apache/openwhisk/core/entity/ActivationResult.scala#L207

  * POST /run returns a response with status != OK (200), response is not truncated but is no valid JSON. We assume that the managed action runtime detected a problem during / after run and and sent a broken response due to a bug.
  * Should not happen for managed runtimes because such failures should be caught in test. Blackbox containers may return improper responses. Can malicious actions cause sending of a broken response?
  * Example: no known examples.
  * Expectation: managed action runtimes should produce sentinels in this situation.

* https://github.com/apache/openwhisk/blob/9bef49fcd47f7922e4226ff7a97385f1313f7d64/common/scala/src/main/scala/org/apache/openwhisk/core/entity/ActivationResult.scala#L211

  * POST /run returns a response with status != OK (200) and response is truncated because it exceeds the size limit (1 MiB). We assume that the managed action runtime detected a problem during / after run and and sent a broken response due to a bug.
  * Should not happen for managed runtimes because such failures should be caught in test. Blackbox containers may return improper responses. Can malicious actions cause sending of a broken response?
  * Example: no known examples.
  * Expectation: managed action runtimes should produce sentinels in this situation.

* https://github.com/apache/openwhisk/blob/9bef49fcd47f7922e4226ff7a97385f1313f7d64/common/scala/src/main/scala/org/apache/openwhisk/core/entity/ActivationResult.scala#L215

  * The TCP connection to the action container "fails" or breaks down during POST /run or while waiting for a response AND the OOM check is positive. We assume that the managed action runtime did not have a chance to react on this error situation.
  * Example: action code consumes too much physical memory by creating too many objects.
  * Expectation: managed action runtimes won't produce sentinels in this situation.

* https://github.com/apache/openwhisk/blob/9bef49fcd47f7922e4226ff7a97385f1313f7d64/common/scala/src/main/scala/org/apache/openwhisk/core/entity/ActivationResult.scala#L219

  * The TCP connection to the action container "fails" or breaks down during POST /run or while waiting for a response AND the OOM check is negative. We assume that the managed action runtime did not have a chance to react on this error situation.
  * Example: the action code breaks the managed action runtime process by exiting.
  * Expectation: managed action runtimes won't produce sentinels in this situation. Malicious actions may be able to delay / prevent sentinel production.

markusthoemmes commented on core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/docker/DockerContainer.scala:
```diff
@@ -257,17 +257,43 @@ class DockerContainer(protected val id: ContainerId,
+   * There are two possible modes controlled by parameter waitForSentinel:
    *
-   * Only parses and returns as much logs as fit in the passed log limit.
+   * 1. Wait for sentinel: tail container log file until two sentinel markers show up. Complete
```
Please break the description to the next line. Indent by one or two spaces if you feel the necessity there.

sven-lange-last replied:
Makes sense. Re-formatting as suggested.

markusthoemmes commented on common/scala/src/main/scala/org/apache/openwhisk/core/containerpool/logging/DockerToActivationLogStore.scala:
```diff
@@ -87,16 +97,18 @@ class DockerToActivationLogStore(system: ActorSystem) extends LogStore {
       logs.concat(
-        Source.single(ByteString(LogLine(Instant.now.toString, "stderr", Messages.logFailure).toJson.compactPrint)))
+        Source.single(
+          ByteString(LogLine(Instant.now.toString, "stderr", Messages.logWarningDeveloperError).toJson.compactPrint)))
```
I'm assuming the log will end with both a logging error and this error, is that correct? If so: maybe we should drop the TimeoutException from the recover below and add that to the the stream here. We can then decide if we want to append the general log failure or this new one.

sven-lange-last replied:
> I'm assuming the log will end with both a logging error and this error, is that correct?

I think no, because:

* If we have a developer error, we won't perform tailing log file reading but only read to the file's current end. In addition, we do not require to find 2 sentinel markers.
* With these conditions, the likelihood to run into idle / completion timeouts should be low.
* All other errors (framing, log limit exceeded) could possibly occur - but there is no correlation to the developer error.

My understanding is that you are thinking about breaking ties when we have two different error conditions like developer error AND timeout / framing error / log limit exceeded?

From my point of view, there are three perspectives on that:

1. As action developer, I use the action logs to understand and debug problems with init'ing and running actions. For the action developer, it probably doesn't matter which error appears in the last log line or whether the two last log lines mention more specific error conditions (`There was an issue ...` and `The action did not initialize or run as expected...`) - as long as there is a hint what happened and that the log is probably truncated.

2. From a container proxy perspective, I want to remove (and not re-use) an action container that had a developer or log collecting and processing error. From that perspective, it doesn't matter which of these two error conditions arrives in the container proxy.

3. As operator, I do not want see indications (system log messages, metrics) for failed log collections if there is a developer error - but log collecting and processing works properly. Only if a timeout / framing error / log limit exceeded occurs, I want a log collecting failure. That's why I excluded `logWarningDeveloperError` as possible error message.

The current variant would not "see" any of the current error messages `There was an issue while collecting your logs. Data might be missing.` and `Logs were truncated because the total bytes size exceeds the limit of ${limit.toBytes} bytes.` in case of a developer error because the last message would always be `The action did not initialize or run as expected. Log data might be missing.`. From that perspective, it would be good append the developer error message after checking the last log line for log collecting / processing errors.

markusthoemmes replied:
Thanks for the exhaustive writeup. My only concern was having multiple error messages in the log file. The missing piece in my head was that we don't tail in the non-sentinel case (I'm hitting my past self for this, I should have introduced a separate path for that which doesn't have the idleTimeout for instance 😂 )

sven-lange-last replied:
@markusthoemmes I pushed a new commit that should properly deal with log collecting errors when there is a developer error. Please have a look. It's no "nice" code though...

markusthoemmes commented on core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/docker/DockerContainer.scala:
```diff
@@ -257,17 +257,43 @@ class DockerContainer(protected val id: ContainerId,
+    // Define a time limit for collecting and processing the logs of a single activation.
+    // If this time limit is exceeded, log processing is stopped and declared unsuccessful.
+    // Allow 2 seconds per MB log limit, with a minimum of 10 seconds.
+    val logProcessingTimeout = limit.toMB.toInt.max(5) * 2.seconds
```
This is somewhat arbitrary and might need some tweaking. Can you run this code through a loadtest to determine we don't hit this overall timeout under load too often?

sven-lange-last replied:
Good point. I checked production histogram metrics for `invoker.collect_logs_finish` and the max value for around 20 million action container log reads is below 1.2 seconds. So I conclude that 20 seconds for a 10 MiB log limit should be a reasonable setting.

markusthoemmes replied:
Per @rabbah's comment below, I'm fine with rather going longer than too short on this for now. It's only a safeguard anyway.

rabbah replied:
nice @sven-lange-last thanks for checking.

sven-lange-last replied:
@markusthoemmes @rabbah I explicitly want to check with you on the total timeout setting. Are you ok with the setting above (2s / MiB log limit, minimum 10s) - given that this is "just" a safeguard to address other currently unknown problems? As discussed, the developer error is mainly addressed by changing from sentinel-based tailing to non-sentinel reading to end-of-file without tailing.

markusthoemmes replied:
Yes I am fine with that.

rabbah commented on core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/docker/DockerContainer.scala:
```diff
@@ -281,9 +309,11 @@ class DockerContainer(protected val id: ContainerId,
       // between two loglines appear without relying on the log frequency in the action itself.
       .idleTimeout(waitForLogs)
+      // Apply an overall time limit for this log collecting and processing stream.
+      .completionTimeout(logProcessingTimeout)
```
It looks to me like this is the crux of the patch - limit the log extraction until sentinel is found. I don’t know what a reasonable rate should be but up to 2s/1MB strikes me as long. I like Markus’ suggestion and wonder what the drain rate might be under load. That’s especially when waiting is on the critical path.

markusthoemmes replied:
It's not really the crux but an additional safeguard as I read it. It's a "catchall".

The crux here is to ignore sentinels (and thus don't tail logs, which guarantees a finish of the stream) in error cases.

rabbah replied:
See my comment above on why I thought/think this is the essence of the patch. Arguably this limit should have already existed when waiting for logs to drain. But I understand the combined effect as well now, thanks to the excellent comments.

sven-lange-last commented on common/scala/src/main/scala/org/apache/openwhisk/core/containerpool/logging/DockerToActivationFileLogStore.scala:
```diff
@@ -123,8 +122,9 @@ class DockerToActivationFileLogStore(system: ActorSystem, destinationDirectory:
-    val isTimedoutActivation = activation.isTimedoutActivation
-    val logs = logStream(transid, container, action, isTimedoutActivation)
+    val logLimit = action.limits.logs
+    val isDeveloperError = activation.response.isContainerError // container error means developer error
```
Yes, that's true - it will only read until the end of the log file and not tail until a sentinel log marker is found. And also for more conditions - see here: https://github.com/apache/openwhisk/pull/4582#issuecomment-520706643.

sven-lange-last commented on common/scala/src/main/scala/org/apache/openwhisk/core/containerpool/logging/DockerToActivationLogStore.scala:
```diff
@@ -75,8 +74,19 @@ class DockerToActivationLogStore(system: ActorSystem) extends LogStore {
+   * in a managed action runtime and prevent the production of sentinels. In that case, log file
+   * reading may continue endlessly.
+   *
+   * For these reasons, do not wait for sentinels to appear in log output when activations end up
```
My goal was to quickly complete log collecting if a developer error occurred. That's accomplished with not tailing the log file on developer error.

The total log collecting timeout was "just" meant as a safeguard for further, currently unknown flaws.

If we used a total log collecting timeout also for developer errors, we would give action containers some additional seconds of CPU time.

sven-lange-last commented on common/scala/src/main/scala/org/apache/openwhisk/core/entity/ActivationResult.scala:
```diff
@@ -55,18 +55,19 @@ protected[core] object ActivationResponse extends DefaultJsonProtocol {
 
   protected[core] def messageForCode(code: Int) = {
-    require(code >= 0 && code <= 3)
+    require(code >= Success && code <= WhiskError)
```
I agree - that also came into my mind... But I was already happy to replace numbers with `val`s.

markusthoemmes commented on common/scala/src/main/scala/org/apache/openwhisk/core/containerpool/logging/DockerToActivationLogStore.scala:
```diff
@@ -113,29 +113,61 @@ class DockerToActivationLogStore(system: ActorSystem) extends LogStore {
+   * @param isDeveloperError did activation fail due to developer error?
+   * @return true if log collecting failed, false otherwise
+   */
+  protected def isLogCollectingError(actLogs: ActivationLogs,
```
The logic here kinda what I was fearing earlier 😂 . Should we at least leave a TODO to return errors from the log gathering itself and append the appropriate lines based on these after the fact?

sven-lange-last replied:
Well... I feel ashamed about this piece of code... at the same time, I had the feeling that I would have to improve other areas as well.

For example: the log reading pipeline recovers exceptions with regular stream elements, i.e. log messages that we check for later on. If an action just logged one of these messages as last log line, we would detect this a log collecting error...

I tried a lot of different approaches and one of the problems was that the `DockerToActivationFileLogStore` has a fan-out to different sinks so that the "injection" of error messages needs to be performed at a reasonable stage of the pipeline. That probably requires a custom flow or sink.

I guess it's a good idea to leave a TODO in the comments.

sven-lange-last commented:
@markusthoemmes @rabbah I think I addressed your review comments. Please check and let me know.

@rabbah concerning your comment / request:

> I think this is a dilation factor we can move to the configuration parameters.

I didn't fully expose this factor as a configuration parameter yet because we were fine so far with also having other timeouts as `val`s local to the class. See https://github.com/apache/openwhisk/blob/b5b88ece0ae08c87948796b38ab18e55ab4b70a7/core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/docker/DockerContainer.scala#L176-L178

* That's why I renamed one of the timeouts for more clarity and added `logCollectingTimeoutPerMBLogLimit` to this list.
* I suggest that we should expose all or none of these config parameters and only if there is a need to change these defaults. @rabbah okay to delay loading these config parameters or do you need config loading already now?

rabbah commented:
>  @rabbah okay to delay loading these config parameters or do you need config loading already now?

Yes, as I wrote, looked good to me already ;)

sven-lange-last commented:
@markusthoemmes please let me know whether anything is still required from your side.

@style95 please check whether you are ok.

I would like to get this PR merged if everybody is ok...
